src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -57,39 +51,40 @@ export class TerminalService implements ITerminalService {
 	private _terminalAltBufferActiveContextKey: IContextKey<boolean>;
 	private _terminalGroups: ITerminalGroup[] = [];
 	private _backgroundedTerminalInstances: ITerminalInstance[] = [];
-	private get _terminalInstances(): ITerminalInstance[] {
-		return this._terminalGroups.reduce((p, c) => p.concat(c.terminalInstances), <ITerminalInstance[]>[]);
-	}
 	private _findState: FindReplaceState;
 	private _extHostsReady: { [authority: string]: IExtHostReadyEntry | undefined } = {};
 	private _activeGroupIndex: number;
+	private _activeInstanceIndex: number;
 	private _linkProviders: Set<ITerminalExternalLinkProvider> = new Set();
 	private _linkProviderDisposables: Map<ITerminalExternalLinkProvider, IDisposable[]> = new Map();
 	private _processSupportContextKey: IContextKey<boolean>;
-
-	public get activeTabIndex(): number { return this._activeGroupIndex; }
-	public get terminalInstances(): ITerminalInstance[] { return this._terminalInstances; }
-	public get terminalGroups(): ITerminalGroup[] { return this._terminalGroups; }
-	public get isProcessSupportRegistered(): boolean { return !!this._processSupportContextKey.get(); }
-
+	private readonly _localTerminalService?: ILocalTerminalService;
+	private readonly _offProcessTerminalService?: IOffProcessTerminalService;
```

Comment:
Don't need to solve this within this PR, but I started calling this the "Primary" service somewhere, taking cues from the "primary extension host". This could use some thinking around naming since we have ITerminalService, ILocal/RemoteTerminalService and ILocal/RemotePtyService - "primary terminal service" is ambiguous, "primary off process terminal service" is too verbose.

Reply:
Yes I agree